internal/consensus/reactor.go

```diff
@@ -1232,12 +1246,16 @@ func (ps *PeerState) PickVoteToSend(votes types.VoteSetReader) (vote *types.Vote
 
 	psVotes := ps.getVoteBitArray(height, round, votesType)
 	if psVotes == nil {
-		return nil, false // Not something worth sending
+		return nil // Not something worth sending
 	}
 	if index, ok := votes.BitArray().Sub(psVotes).PickRandom(); ok {
-		return votes.GetByIndex(int32(index)), true
+		vote := votes.GetByIndex(int32(index))
```

Comment:
even though you're passing an index that should be in the voteSet, in the `GetByIndex` function there's no bound checks and it will return `nil` only if the `voteSet` is nil. Adding a defensive logic in the `GetByIndex` such as the one below can improve the logic

```
if valIndex >= 0 && valIndex < len(voteSet.votes) {
    return voteSet.votes[valIndex]
} else {
   return nil
}
```

right now it just returns `return voteSet.votes[valIndex]`. The logic seems OK everywhere it's invoked because you always check for `ok` and the index should be inside the boundaries but since this is an exported function, good idea to validate the boundaries of the slice 😉

Reply:
This is a good point, but if no one changes `votes` between the two calls, how can we get an invalid index? Actually, a bad index would only be produced if the size of the validator set changes, which can only happen over heights.

Reply:
Closing this ATM as there seems not to be away to expose the problem. @andynog feel free to open an issue if you would like to take a closer look at this